server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -215,9 +215,12 @@ public void restoreSnapshot(final RestoreSnapshotRequest request, final ActionLi
                     dataStreams = new HashMap<>();
                 } else {
                     globalMetadata = repository.getSnapshotGlobalMetadata(snapshotId);
-                    dataStreams = globalMetadata.dataStreams();
-                    if (request.includeGlobalState() == false) {
-                        dataStreams.keySet().retainAll(requestedDataStreams);
+                    final Map<String, DataStream> dataStreamsInSnapshot = globalMetadata.dataStreams();
```

Comment:
1. We gotta only restore those data streams actually matched by the restore request, even if we include global state.
2. We really shouldn't be mutating the deserialization result here. Come to think of it, we also shouldn't be deserializing the data streams to a mutable map, that invites all kinds of weird bugs. I'll open a PR to fix that.

Reply:
> Come to think of it, we also shouldn't be deserializing the data streams to a mutable map, that invites all kinds of weird bugs.

I don't fully understand. Do you mean that the `DataStream` instances also need to be copied/cloned? So that the deserialization result cannot be altered at all?

Reply:
Aren't the `DataStream` instances immutable? If they aren't already I think we should make them immutable.

Reply:
In practice they are (all operations that modify a data stream create a new instance), however in order to make it stricter we should wrap the `indices` list always with `Collections.unmodifiableList(...)`.